substrate-node/pallets/pallet-tfgrid/src/node.rs

```diff
@@ -0,0 +1,211 @@
+    /// minimum or exceeds the maximum allowed length or contains invalid ASCII
+    /// characters.
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        ensure!(
+            value.len() >= MIN_COUNTRY_NAME_LENGTH.saturated_into(),
+            Self::Error::NodeCountryNameTooShort
+        );
+        let bounded_vec: BoundedVec<u8, ConstU32<MAX_COUNTRY_NAME_LENGTH>> =
+            BoundedVec::try_from(value).map_err(|_| Self::Error::NodeCountryNameTooLong)?;
+        ensure!(
+            validate_country_name(&bounded_vec),
+            Self::Error::InvalidNodeCountryName
+        );
+        Ok(Self(bounded_vec, PhantomData))
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> PartialEq for CountryName<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> Clone for CountryName<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), self.1)
+    }
+}
+
+fn validate_country_name(input: &[u8]) -> bool {
+    // TODO: find better alternative
+    input
+        .iter()
+        .all(|c| c.is_ascii_alphabetic() || c.is_ascii_whitespace())
+}
+
+/// A city name (ASCI Characters).
+#[derive(Encode, Decode, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+#[codec(mel_bound())]
+pub struct CityName<T: Config>(
+    pub(crate) BoundedVec<u8, ConstU32<MAX_CITY_NAME_LENGTH>>,
+    PhantomData<(T, ConstU32<MAX_CITY_NAME_LENGTH>)>,
+);
+
+impl<T: Config> TryFrom<Vec<u8>> for CityName<T> {
+    type Error = Error<T>;
+
+    /// Fallible initialization from a provided byte vector if it is below the
+    /// minimum or exceeds the maximum allowed length or contains invalid ASCII
+    /// characters.
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        ensure!(
+            value.len() >= MIN_CITY_NAME_LENGTH.saturated_into(),
+            Self::Error::NodeCityNameTooShort
+        );
+        let bounded_vec: BoundedVec<u8, ConstU32<MAX_CITY_NAME_LENGTH>> =
+            BoundedVec::try_from(value).map_err(|_| Self::Error::NodeCityNameTooLong)?;
+        ensure!(
+            validate_city_name(&bounded_vec),
+            Self::Error::InvalidNodeCityName
+        );
+        Ok(Self(bounded_vec, PhantomData))
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> PartialEq for CityName<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> Clone for CityName<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), self.1)
+    }
+}
+
+fn validate_city_name(input: &[u8]) -> bool {
+    // TODO: find better alternative
+    input
+        .iter()
+        .all(|c| c.is_ascii_alphabetic() || c.is_ascii_whitespace())
```

Comment:
Maybe we should have a list of all valid "Countries" and cross check the input with the list values

see: https://raw.githubusercontent.com/shivammathur/countrycity/master/data/geo.json

Reply:
Let s do it!